rfcs/20200920-keras-nlp-bert.md

```diff
@@ -0,0 +1,379 @@
+      bias_regularizer: Regularizer for dense layer biases.
+      activity_regularizer: Regularizer for dense layer activity.
+      kernel_constraint: Constraint for dense layer kernels.
+      bias_constraint: Constraint for dense layer kernels.
+      use_bias: Whether to enable use_bias in attention layer. If set False,
+        use_bias in attention layer is disabled.
+      norm_first: Whether to normalize inputs to attention and intermediate
+        dense layers. If set False, output of attention and intermediate dense
+        layers is normalized.
+      norm_epsilon: Epsilon value to initialize normalization layers.
+      output_dropout: Dropout probability for the post-attention and output
+        dropout.
+      attention_dropout: Dropout probability for within the attention layer.
+      inner_dropout: Dropout probability for the first Dense layer in a
+        two-layer feedforward network.
+      attention_initializer: Initializer for kernels of attention layers. If set
+        `None`, attention layers use kernel_initializer as initializer for
+        kernel.
+      **kwargs: keyword arguments/
+    """
+```
+
+### Layers -- SelfAttentionMask
+
+```python
+class SelfAttentionMask(tf.keras.layers.Layer):
+  """Create 3D attention mask from a 2D tensor mask.
+
+    inputs[0]: from_tensor: 2D or 3D Tensor of shape
+      [batch_size, from_seq_length, ...].
+    inputs[1]: to_mask: int32 Tensor of shape [batch_size, to_seq_length].
+
+    Returns:
+      float Tensor of shape [batch_size, from_seq_length, to_seq_length].
+  """
+
+  def call(self, inputs, to_mask):
+```
+
+### Layers -- OnDeviceEmbedding
+
+```python
+class OnDeviceEmbedding(tf.keras.layers.Layer):
+  """Performs an embedding lookup suitable for accelerator devices.
+
+  This layer uses either tf.gather or tf.one_hot to translate integer indices to
+  float embeddings.
+
+  Arguments:
+    vocab_size: Number of elements in the vocabulary.
+    embedding_width: Output size of the embedding layer.
+    initializer: The initializer to use for the embedding weights. Defaults to
+      "glorot_uniform".
+    use_one_hot: Whether to use tf.one_hot over tf.gather for the embedding
+      lookup. Defaults to False (that is, using tf.gather). Setting this option
+      to True may improve performance, especially on small vocabulary sizes, but
+      will generally require more memory.
+    scale_factor: Whether to scale the output embeddings. Defaults to None (that
+      is, not to scale). Setting this option to a float will let values in
+      output embeddings multiplied by scale_factor.
+  """
+
+  def __init__(self,
+               vocab_size,
+               embedding_width,
+               initializer="glorot_uniform",
+               use_one_hot=False,
+               scale_factor=None,
+               **kwargs):
+```
+
+### Layers -- PositionEmbedding
+
+```python
+class PositionEmbedding(tf.keras.layers.Layer):
```

Comment:
There's more than one type of positional embedding in the literature (some people want them periodic, for example) so I think it's a good idea to call it something more specific.

Reply:
@tanzhenyu The other common encoding is relative position encoding: https://github.com/tensorflow/models/blob/master/official/nlp/modeling/layers/position_embedding.py#L26
Note that, we need to consolidate the implementation related to: https://github.com/tensorflow/tensor2tensor/pull/177
TransformerXL implemented the paper version: https://github.com/tensorflow/models/blob/master/official/nlp/modeling/networks/xlnet_base.py#L341 

Thus, for the layer naming, maybe: PositionEmbedding->LearnedPositionEmbedding and the relative one is RelativePositionEmbedding/Encoding?

Reply:
Hmmm...PositionEmbedding and PositionalEncoding are two different things. I believe when we say `Embedding`, it should always imply it's trainable, on the other hand `Encoding` implies it's not trainable. Thoughts?